### DISCUSSION TITLE
MacOS support

### DISCUSSION THREAD

**Discussion by Schlaubischlump** [BLeeEZ/amperfy#297] [Show and tell]
Hi @BLeeEZ, 

I really like your app! Great work :) 

I thought I could help out a little bit and add macOS Catalyst support to Amperfy. Most things are working (reordering sidebar is not, I disabled it for now). I only tested it on macOS 14.5. If you are interest, I will cleanup the code and create a pull request in the next couple of days / weeks. 

I attached some screenshots to give you an idea how the app looks like. 

![Screenshot 2024-07-20 at 23 47 52](https://github.com/user-attachments/assets/5a854734-e4ea-4df2-b363-0524d84ef044)
![Screenshot 2024-07-21 at 00 28 48](https://github.com/user-attachments/assets/c5f1426e-2e45-41a5-9f2b-0330285a653b)
<img width="1136" alt="Screenshot 2024-07-21 at 00 28 55" src="https://github.com/user-attachments/assets/0aa18c93-1618-43ee-9579-a646c2ae7277">
![Screenshot 2024-07-21 at 00 24 30](https://github.com/user-attachments/assets/b4286836-0465-4b01-acba-1e57c674aeab)
![Screenshot 2024-07-21 at 00 23 40](https://github.com/user-attachments/assets/b80330e0-4460-44f6-8080-da15b92369e9)
![Screenshot 2024-07-21 at 00 21 26](https://github.com/user-attachments/assets/3f61757f-a189-4f69-b17e-bafb71756e27)
![Screenshot 2024-07-21 at 00 21 12](https://github.com/user-attachments/assets/4368e3ed-a988-4e74-a2d9-50ae2c800b99)
![Screenshot 2024-07-21 at 00 22 41](https://github.com/user-attachments/assets/97e90178-fcc3-46f2-a465-fe0f67271162)

**Comment by BLeeEZ** [comment]
Hi @Schlaubischlump,
thank you. The screenshots look great. From my side adding MacOS Catalyst support is a great idea.
Thank you for the work you already did.
I am looking forward for the PR :)

**Comment by BLeeEZ** [comment]
@Schlaubischlump @rudolphmax
Here is a list of topics for Mac that could be done if someone is interrested:
- create an own "menu bar" for different useful actions (player action, visibility, OfflineMode?)
- add a special embedded Player view to the main window
- rework the top part of window: smaller play&shuffle buttons, display song count in navbar?,  make "..." button look nicer: <img width="877" alt="image" src="https://github.com/user-attachments/assets/09ef5105-514c-479c-9caf-92389ee8372a">
- reduce table view rows sizes and make them look more like a row in a desktop application
- use NSTableView instead of UITableView to make use of resizeable columns
- improve/simplify tab handling (menu bar->View->Show Tab Bar): <img width="1309" alt="image" src="https://github.com/user-attachments/assets/06c169eb-11e0-47fc-9a48-3964653b29da">

I am also interested in your opinions on that topics.

**Comment by Schlaubischlump** [comment]
Nice work on the pull request and thanks for merging it so quickly! Some thoughts on the points you mentioned:

> create an own "menu bar" for different useful actions (player action, visibility, OfflineMode?)

> reduce table view rows sizes and make them look more like a row in a desktop application

Sounds like good ideas and shouldn't be too hard to implement I think.

> add a special embedded Player view to the main window

- I experimented a little bit with this. My idea was to introduce a triple column layout on macOS where the right hand side is just the `LargeNowPlayingViewController`. One could also only add the queue or the lyrics to the third column and keep the player in the navigationbar or a toolbar on the bottom. However, the usage of the storyboard made it really difficult to change the layout from two column to three columns for one particular platform. There is just no option for that. I just ended up forcing three columns on all platforms to keep on experimenting, but the `SplitVC` code got really confusing with the third column. That's why I ultimately abandoned the idea.

> rework the top part of window: smaller play&shuffle buttons, display song count in navbar?, make "..." button look nicer:

- Is there a particular reason why we use the `.pad` style for the `UINavigationBar`  in `BasicCollectionViewController`? I personally prefer the look of the `.mac` style more, although I have to admit that this is a matter of personal taste. Currently, it feels inconsistent with the Search and Settings tabs. 

<img width="683" alt="Screenshot 2024-08-06 at 23 32 59" src="https://github.com/user-attachments/assets/3a7567dd-921d-433a-ab50-5d99ebfae334">

> use NSTableView instead of UITableView to make use of resizeable columns

- I'm not sure that this is the best way to approach this. Mixing UIKit and AppKit always feels hacky. I think it would be better, albeit being more effort, to recreate a multicolumn tableview in UIKit that mimics Apples `Table` design found in SwiftUI ([example](https://useyourloaf.com/blog/swiftui-tables-quick-guide/)). That way, iPadOS could also profit from the multicolumn layout. 

> improve/simplify tab handling (menu bar->View->Show Tab Bar):

Do you have any particular ideas in mind? Maybe a right click menu to `open in new tab`? 
It really annoys me that the sidebar moves down when the tabbar is displayed. I think there might be a workaround by setting the navigationbar of the SideVC to translucent, hack the navbar frame to get out of the way and inset the top of the collectionView by some negative value based on the tabbar height. I played around with this, but I always end up with an ugly, unmaintainable mess that is only working half of the time.

**Comment by BLeeEZ** [comment]
> add a special embedded Player view to the main window

I was more thinking about something like that: <img width="1041" alt="image" src="https://github.com/user-attachments/assets/c1ec33d5-6631-4de4-8f87-2ab4f2321f1c">
A complete player interaction that is available from every library/search/setting view.
The ideas/investigations you mentioned sound good too. I like the idea with the lyrics panel.


> Is there a particular reason why we use the .pad style for the UINavigationBar

I prefer the look of .mac too. I did this change mainly because the navbar-Buttons can't be updated in .mac style. I display a checkmark in the sort category beside the currently active element. The button update only works in .pad style. Either we remove the checkmark on Catalyst and use .mac or you have an idea to overcome this problem. The more I think about that it, the more I prefer the .mac way and simply remove the checkmarks.

> use NSTableView instead of UITableView to make use of resizeable columns

I did some further reading and I agree that NSTableView is not the way we should go. I like the idea with SwiftUI at least for iPad and macOS. I did some tests in the past to convert the complete App to SwiftUI but I was limited by the functionality SwiftUI provided then (Swipe on list items was not supported). Now some important features are available and on iPad/macOS swiping a song is not that important in SongsVC. As first step implementing SongsVC for iPad/macOS in SwiftUI with your provided example link sounds good. 

> improve/simplify tab handling (menu bar->View->Show Tab Bar)

Currently the user must activate the tabbar once. I think this should be activated by default. The tab names are showing the sceneTitle instead of NavBar title. You are right: the sidebar moves down when tab bar is active, I didn't noticed that. If the user doesn't like that, there should still be an option to hide tab bar. Maybe we can use the navbar in the sidebar to place some useful buttons or a search field in there: <img width="209" alt="image" src="https://github.com/user-attachments/assets/75039280-1570-4871-a1ae-2aa3220d5508">

**Comment by Schlaubischlump** [comment]
> I prefer the look of .mac too. I did this change mainly because the navbar-Buttons can't be updated in .mac style. I display a checkmark in the sort category beside the currently active element. The button update only works in .pad style. Either we remove the checkmark on Catalyst and use .mac or you have an idea to overcome this problem. The more I think about that it, the more I prefer the .mac way and simply remove the checkmarks.

- Ah I see. There is an easy fix for that. Catalyst doesn't like it when you try to exchange a button in the navigationbar with a new one (Don't ask me why). However, you can just update the existing bar button. 

E.g in ArtistVC line 120 you have the following code: 

```
optionsButton = OptionsBarButton() // remove this
optionsButton.menu = UIMenu(children: actions)
navigationItem.rightBarButtonItems = [optionsButton]
```

Just remove the first line and instead initialize the button when you define the attribute: 

```
private var optionsButton: UIBarButtonItem = OptionsBarButton()
```

I'll create a pull request. Edit: Done


> Currently the user must activate the tabbar once. I think this should be activated by default. The tab names are showing the sceneTitle instead of NavBar title. You are right: the sidebar moves down when tab bar is active, I didn't noticed that. If the user doesn't like that, there should still be an option to hide tab bar. Maybe we can use the navbar in the sidebar to place some useful buttons or a search 

- I couldn't figure out how to place anything in that area... right or leftbarbuttonitems are always added to the navigationBar of the detail view. The titleView property is ignored. A UISearchController on the navigationItem is added below the area and moves up and down as well. Changing the navigationBar to inline instead of stacked will remove the searchcontroller completely.

**Comment by Schlaubischlump** [comment]
I worked on a small update for macOS, which I think looks cleaner.

The settings are now separated into a new window which can be accessed using the menubar.
I reworked the sidebar to remove the settings and added a new searchbar instead of the search icon. The searchbar does not move when the tabbar is display, but does scroll with the sidebar (same behaviour as the native music app). The scope of the searchbar is limited to the currently active tab.

<img width="226" alt="Screenshot 2024-08-12 at 00 45 32" src="https://github.com/user-attachments/assets/735aa3ef-21ce-4de4-a5b2-b8c0cc4e5706">
<img width="1136" alt="Screenshot 2024-08-12 at 00 45 43" src="https://github.com/user-attachments/assets/2973d59a-4ac8-43f7-9286-8fbaadf69056">
<img width="1136" alt="Screenshot 2024-08-12 at 01 18 19" src="https://github.com/user-attachments/assets/ec9b411b-18cb-4a81-90e4-9303496c075a">

**Comment by BLeeEZ** [comment]
Cool 👍 This looks really great.
Only one small question: is it possible to move the version/build info, offline mode and screen lock to a new subsection. It looks a little bit cluttered in the settings sidebar. If it is too much work for now, we can submit that in a separate commit/PR.

**Reply by Schlaubischlump** [reply]
Good point. I was already thinking about reorganizing the settings to make it look more mac like. I had something like this in mind: ![image](https://github.com/user-attachments/assets/40f233ac-29ac-408c-b6fc-8f19f0e5bd15)

Basically, move everything in the sidebar to a tab and add an additional "General" tab for the info you just mentioned.

**Reply by BLeeEZ** [reply]
Perfect 👍

**Reply by Schlaubischlump** [reply]
I'm getting there. 

<img width="752" alt="Screenshot 2024-08-14 at 15 23 47" src="https://github.com/user-attachments/assets/92fb7042-66c3-405e-9b87-ab3ea968c9bc">

It would be nice if the menu would look more like the screenshot below. In particular, a ":" follows after each text on the left-hand side, the items are center aligned, the list has no row background etc. But I think to get all these details right I either need to rewrite all setting views for Catalyst or include a lot of ifdefs in the existing files. 

<img width="942" alt="Screenshot 2024-08-14 at 15 25 46" src="https://github.com/user-attachments/assets/244c08e2-0958-448f-a310-3092e9871dd4">

**Comment by BLeeEZ** [comment]
I like your version. The version from the below screenshot is not needed from my point of view.

**Comment by Schlaubischlump** [comment]
Hi @BLeeEZ, 

I'm currently working on better player controls. I attached a screenshot of my current progress. All UI elements are already working. My plan for the next pull request is to completely replace the ugly bottom bar with items in the navigation bar. In a next step, I want to reintroduce lyrics and the music queue as a third column. There are just a couple of design questions I have. Maybe you have some ideas. 

1. The toolbar overflow menu on macOS is broken. If you click on it nothing happens. That is, we should make sure that all navigation bar items are always visible. Thats why I think it would make sense to remove the searchbars from all view controller and instead use the space for the music controls. I think that is reasonable, since you can still search with a single click in the sidebar. You just loose the option to filter the current view. The screenshot below just shows the SearchVC, which does not include a searchbar anyway.

2. Does it make sense to also add a volume slider to the navigation bar like the music app? 

3. What about the sleep timer, playback speed and podcast buttons. It seems to me that putting all of these in the navigation bar makes the UI too cluttered. On top of that we still have the annoying bug with the overflow menu. 

<img width="1361" alt="Screenshot 2024-08-22 at 00 22 24" src="https://github.com/user-attachments/assets/c4b5b4bd-7d11-4b5b-9178-897afa3455a3">

**Comment by BLeeEZ** [comment]
First of all, this looks really good. I am not quite sure what you mean with toolbar overflow menu behaviour.

1. is it possible to move the search bar in the "sub bar" like here: <img width="1217" alt="image" src="https://github.com/user-attachments/assets/51fc5dd9-aa9b-4528-991b-de0bfaee8518">

2. I  think this is not needed (iOS has none too)

3. I think some UI elements can be moved to the toolbar (Music/Podcast, Playback speed, Sleep timer). This should reduce the cluttering: <img width="309" alt="image" src="https://github.com/user-attachments/assets/69a2f27a-f3a3-49e2-ba39-37a9ca7dd261">

**Reply by Schlaubischlump** [reply]
Overflow menu: When you resize a window, such that there is not enough space, then menu items are moved into the "overflow menu". 

![image](https://github.com/user-attachments/assets/fc851636-5512-44bf-918c-f004f08b7d0d)

1. I haven't managed to create a satisfying "subbar" yet, but I'm trying different options. Maybe a second toolbar works, otherwise we could manipulate each view controller separately. I'll keep you updated. 

2. Sounds good. 

3. That sounds like a good idea! Somehow I forgot about the menu bar items. Thanks :)

**Comment by Schlaubischlump** [comment]
I think I figured out a solution that works and requires minimal code changes. We embed every secondary viewController in the splitView in two UINavigationController. The outer NavigationController hosts the media control toolbar (.mac style). The inner NavigationController hosts the same controls as on iOS including the searchbar (.pad style). This is the result: 

<img width="1136" alt="Screenshot 2024-08-26 at 01 55 59" src="https://github.com/user-attachments/assets/bfeac8e2-bf1c-4a57-a043-e058e56fee8a">

I also added a mini player. It is activated the same way as in the music app (clicking on the artwork in the toolbar).

<img width="512" alt="Screenshot 2024-08-26 at 00 19 05" src="https://github.com/user-attachments/assets/e45b95a1-f6dd-4ee9-8879-c90ee3810de7">

"Controls" menu as suggested: 

<img width="226" alt="Screenshot 2024-08-26 at 01 34 43" src="https://github.com/user-attachments/assets/0ea12104-4b5c-4de7-9c9f-7f23f08fd5b9">

What do you think?

**Reply by BLeeEZ** [reply]
Looks good. 👍

**Comment by BLeeEZ** [comment]
Hi @Schlaubischlump,
we are preparing the next Beta. iOS version works just as before. With the macOS version we have a problem.
We are able to create a macOS Bundle but we fail when we try upload it to App Connect with the error message:

```
Asset validation failed
Bad Bundle Executable. You must include a valid CFBundleExecutable key in the nested bundle
AppKitintegration [net.hildebrand-it.amperfy-music.pkg/Payload/Amperfy.app/Contents/Frameworks/
AppKitintegration.framework] property list file. (ID: 6173d62e-b2a6-417a-b19e-9e08b1b2879c)
```

Do have an idea how to fix that. We already tried a lot, but we are not able to get around this error message. We tried the following adjustments to the `AppKitIntegration/Info.plist` file:

- removing `<key>CFBundleExecutable</key> <string>$(EXECUTABLE_NAME)</string>`
- using `<key>CFBundleExecutable</key> <string>AppKitIntegration</string>`
- using `<key>CFBundleExecutable</key> <string>Amperfy</string>`

**Reply by Schlaubischlump** [reply]
Sadly I have never seen this error, but I know that this approach works for integrating an AppKit bundle. steventroughtonsmith provides this wonderful example code of how to integrate an AppKit bundle to implement a settings: 

https://github.com/steventroughtonsmith/CatalystPrefsWindow

I'm pretty sure I messed up the integration of the bundle somewhere. I removed and readded the bundle and pushed the new version to my master branch. Maybe you can check it out and try the new version. My break is over now, but I can have another look in the evening if it doesn't work. 

Edit: Opened a pull request

**Comment by daniel-hiller** [comment]
Hi,
really nice work and thanks for the Mac version.
I'm  running the latest beta on MacOS 15.0.1 (24A348).

Bug:
In the login mask, there is no cursor when selecting a field or when typing, also selecting the text is not working/displayed.

Improvement:
For me it would be cool if songs are just played on double click (maybe as an option in the settings )
